src/arcsf/eval/utils.py

```diff
@@ -0,0 +1,299 @@
+        "forget_tr": forget_dict["truth_ratios"],
+        "rouge_scores": retain_dict["rougeL_recall"],
+    }
+
+
+def get_metrics(
+    base_truth_ratios: torch.Tensor,
+    test_values: dict[torch.Tensor],
+) -> dict[float, float, float, float, float]:
+    """
+    Retrieves metrics for tracking an evaluation.
+
+    Args:
+        base_truth_ratios: base model truth ratios to compare against in the ks_test
+        test_values: recorded values from the evaluation of the test model, including
+        both ks test settings
+
+    Returns:
+        result_dict : dictionary contatining the metrics we would like to track
+    """
+    forget_quality_one_sided = np.log(
+        # the case when the truth ratio CDF for forget is greater than the base model
+        ks_test(test_values["forget_tr"], base_truth_ratios, alternative="greater")
+    )
+    forget_quality_two_sided = np.log(
+        # the general measure of 'closeness' of the truth ratio CDF for forget
+        # against the base model
+        ks_test(test_values["forget_tr"], base_truth_ratios)
+    )
+
+    # need to transform these according to table 1 in the TOFU paper
+    transform_retain_tr = torch.clamp((1 - test_values["retain_tr"]), 0)
+    retain_tr = torch.mean(transform_retain_tr)
+    # calculate other scores
+    rouge_score = torch.mean(test_values["rouge_scores"])
+    model_utilty = hmean([retain_tr, rouge_score])
+
+    result_dict = {
+        "mean_tr_retain": retain_tr.item(),
+        "mean_rouge_score": rouge_score.item(),
+        "forget_quality_1": forget_quality_one_sided.item(),
+        "forget_quality_2": forget_quality_two_sided.item(),
+        "model_utility": model_utilty.item(),
+    }
+
+    return result_dict
+
+
+def all_eval(
+    model: torch.nn.Module,
+    dataset: EvalQADataset,
+    batch_size: int,
+    device: torch.device,
+    tokenizer: transformers.PreTrainedTokenizer,
+    **generate_kwargs: dict,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """Performs quantitative evaluation of the selected model over selected data.
+    Returns the input-wise losses on target answers for the ground truth and perturbed
+    answers.
+
+    Args:
+        model : Transformers model used to perform evaluation on
+        dataset : Dataset to perform evaluation on
+        batch_size : batch size for dataloader
+        device : Pytorch device on which to perform computation
+
+    Returns:
+        output_dict : output dictionary with the values for analysis
+    """
+    # move model to device create dataloader, initialise all_losses tensor
+    model = model.to(device)
+    n_perturbed = dataset.n_perturbed
+    data_loader = DataLoader(dataset, batch_size=batch_size, shuffle=False)
+    output_dict = {
+        "all_losses": torch.zeros(
+            (dataset.__len__(), n_perturbed + 1), dtype=torch.float64
+        ),
+        "truth_ratios": torch.zeros(dataset.__len__()),
+        "rougeL_recall": torch.zeros(dataset.__len__()),
```

Comment:
Use `len` rather than calling `__len__` directly


```suggestion
        "all_losses": torch.zeros(
            (len(dataset), n_perturbed + 1), dtype=torch.float64
        ),
        "truth_ratios": torch.zeros(len(dataset)),
        "rougeL_recall": torch.zeros(len(dataset)),
```

Reply:
made this changem since I was using it three times, I just created a variable. Is there a reason we prefer len() over calling __len__ ?